man/mkpwd.1
.TH mkpwd 1 "October 2004" minitools
.
.SH NAME
mkpwd \- Generate random initial passwords
.
.SH SYNOPSIS
\fBmkpwd\fP [-DHV] [-N \fInum\fP] {-\fIc string\fP} [\fIspec\fP]
.
.SH DESCRIPTION
Generate random passwords according to \fIspec\fP and write them
to standard output. All characters in \fIspec\fP are output verbatim,
except for sequences \fInc\fP where \fIn\fP is a decimal number and
\fIc\fP is a single lower case letter, naming an alphabet.
.PP
There are 26 alphabets, named \fBa\fP through \fBz\fP; all of these
can be changed with the -\fIc string\fP (\fIc\fP in [a-z]) option.
Some alphabets are preloaded:
\fBa\fP is [a-zA-Z],
\fBb\fP is [a-z],
\fBc\fP is [A-Z],
\fBd\fP is [0-9], and
\fBz\fP is [a-zA-Z0-9] without [0O1Il]
because these are easily confused in many typefaces.
Note that \fBmkpwd\fP is not clever enough to understand
the shorthand notations for character ranges used here;
instead, you have to explicitly list all characters in the
alphabet.
.PP
If no \fIspec\fP is given on the command line, \fBmkpwd\fP
uses the value of the environment variable MKPWDSPEC.
If this variable is not defined or its value is the empty string,
then \fBmkpwd\fP uses \fB8z\fP, i.e., it generates 8 random
characters from the alphabet \fBz\fP described above.
.PP
If anything goes wrong, \fBmkpwd\fP complains to standard error.
Exit codes are \fB0\fP for success and \fB127\fP on error.
.
.SH OPTIONS
.TP 5
.BI "-N " num
Generate \fInum\fP passwords instead of just one.
.TP 5
.BI - "c string"
Set alphabet \fIc\fP to \fIstring\fP, \fIc\fP in [a-z].
.TP 5
.B -D
Debug mode: print the loaded alphabets to stderr.
.TP 5
.B -V
Show version to standard output and quit.
.TP 5
.B -H
Show quick help to standard output and quit.
.
.SH EXAMPLES
.nf
.RB "$ " mkpwd
mgH9RJo4
.RB "$ " "mkpwd 8z"
ZCszZKRR
.RB "$ " "mkpwd 6c"
QZZMCA
.RB "$ " "mkpwd -a 01 16a"
1001001111001000
.RB "$ " "mkpwd 3b5d"
zap28565
.RB "$ " "mkpwd -N 5 2d-3d-3d"
10-153-445
12-736-513
16-770-938
36-526-833
02-386-421
.RB "$ " "mkpwd -D -N 0" "  # show preloaded alphabets"
a: abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ
b: abcdefghijklmnopqrstuvwxyz
c: ABCDEFGHIJKLMNOPQRSTUVWXYZ
d: 0123456789
z: abcdefghijkmnopqrstuvwxyzABCDEFGHJKLMNPQRSTUVWXYZ23456789
.fi
.
.SH BUGS
Probably many -- please report.
Certainly, one deficiency is that mkpwd can only generate passwords
of a fixed length. Another one is, that we should allow for an external
source of random data to be specified, such as \fI/dev/random\fP on Linux
systems. This could be passed as an option on the command line.
At present, a simple linear congruential random number generator
is used, seeded with the current time plus the process ID.
.
.SH AUTHOR
Written by UJR in 2004.
.br
Released under the GNU General Public License (GPL).


src/mkpwd.c
/* mkpwd - generate initial random passwords
 * Usage: mkpwd [-VD] [-N num] {-<c><alphabet>} [spec]
 * Want: -R randfile to read random bytes from eg /dev/random
 * History: ujr/2004-10-30 created
 * License: GNU General Public License (GPL)
 */

#include <ctype.h>   /* isdigit */
#include <stdio.h>
#include <stdlib.h>  /* getenv */
#include <string.h>  /* strlen */
#include <time.h>    /* time */
#include <unistd.h>  /* getpid */

#include "common.h"

static char id[] = "mkpwd by ujr/2004-10-30\n";

int identity(void);
int usage(const char *s);
void generate(const char *alph[], const char *spec);
int setalph(int i, char *s);
int getint(const char *s, int *val);
void putr(int n, const char *s);
void putn(const char *s, int n);

void rndseed(unsigned seed);
unsigned rnd(unsigned lo, unsigned hi);

char *me = "mkpwd";
const char *alph[26]; /* the 26 alphabets */

int main(int argc, char **argv)
{
  int c, num = 1, debug = 0;
  const char *spec = 0; /* pwd spec; 0 means "8z" */

  /* Initialise alphabets */
  for (c = 0; c < 26; c++) alph[c] = 0;
  alph[0] = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ";
  alph[1] = "abcdefghijklmnopqrstuvwxyz";
  alph[2] = "ABCDEFGHIJKLMNOPQRSTUVWXYZ";
  alph[3] = "0123456789";
  alph[25] = "abcdefghijkmnopqrstuvwxyzABCDEFGHJKLMNPQRSTUVWXYZ23456789";

  (void) argc; /* unused */
  if (*argv && **argv) me = *argv;

  /* Homebrewn command line processing */
  args: while (*++argv && (**argv == '-')) {
    while ((c = *++argv[0])) switch (c) {
      case 'N': if (getint(*++argv, &num)) goto args;
                return usage("missing argument");
      case 'V': return identity();
      case 'h': return usage(0);
      case 'D': debug = 1; break;
      case '-': argv++; goto endargs;
      default:  if (setalph(c, *++argv)) goto args;
                return usage("invalid option");
    }
  }
endargs:
  if (*argv) spec = *argv++;
  if (*argv) return usage("too many arguments");

  /* If no spec on cmd line, check environ */
  if (!spec) spec = getenv("MKPWDSPEC");
  if (spec && *spec == '\0') spec = 0;

  rndseed((unsigned) time(NULL) + getpid());

  if (debug) for (c = 0; c < 26; c++) if (alph[c])
    fprintf(stderr, "%c: %s\n", c+'a', alph[c]); /* XXX */

  while (num--) generate(alph, spec);

  return SUCCESS;
}

int identity(void)
{
  return fputs(id, stdout) >= 0 ? SUCCESS : FAILHARD;
}

int usage(const char *errmsg)
{
  const char *args = "[-VD] [-N num] {-<c> alphabet} [spec]";
  if (errmsg) {
    fprintf(stderr, "%s: %s\n", me, errmsg);
    fprintf(stderr, "Usage: %s %s\n", me, args);
  }
  else {
    fprintf(stdout, "Generate passwords according to spec; see manual.\n");
    fprintf(stdout, "Usage: %s %s\n", me, args);
  }
  return errmsg ? FAILHARD : SUCCESS;
}

void generate(const char *alph[], const char *spec)
{
  if (spec) while (*spec) {
    if (isdigit(*spec)) { int c, n;
      c = getint(spec, &n);
      if (c && n && islower(spec[c]))
        putr(n, alph[spec[c]-'a']), spec++;
      else putn(spec, c);
      spec += c;
    }
    else putc(*spec++, stdout);
  }
  else putr(8, alph[25]); /* default spec */

  putc('\n', stdout);
}

int getint(const char *s, int *val)
{
  register int c, i = 0;
  register const char *p = s;

  if (s) while ((c = (unsigned char) (*p - '0')) < 10)
    i = 10 * i + c, p++;
  if (p > s && val) *val = i;
  return p - s; /* #chars scanned */
}

int setalph(int i, char *s)
{
  if (i < 'a' || i > 'z') return 0; /* error */
  alph[i - 'a'] = (s && *s) ? s : 0; /* map empty to null */
  return 1; /* OK */
}

void putr(int n, const char *a)
{
  int alen;
  if (!a || !*a) return;
  alen = strlen(a) - 1;
  while (n-- > 0) putc(a[rnd(0,alen)], stdout);
}

void putn(const char *s, int n)
{
  if (s && n > 0) fwrite(s, 1, n, stdout);
}

/* Quick+Dirty random numbers
 *
 * Based on chapter 7.1 in Numerical Recipes in C (www.nr.com).
 * It is a linear congruential generator with the constants m (modulus),
 * a (multiplier), and c (increment) taken from the table on page 285.
 *
 * TODO Find better constants for 8bit randoms; adjust types.
 */
static unsigned long r = 1;
void rndseed(unsigned seed) { r = seed; }

unsigned rnd(unsigned lo, unsigned hi)
{
  static unsigned a=9301, c=49297, m=233280;
  r = (r * a + c) % m;
  return lo + ((hi-lo+1)*r)/m;
}
